### ISSUE TITLE
Allow multiple connections from same client IP

### ISSUE THREAD

**Issue by benfrew13** [transmission/transmission#4594]
### What is the issue?

Transmission appears to ignore multiple peers with the same IP address, only serving them one-by-one. I assume this is in place to better manage leechers and not waste bandwidth, but I've recently found myself in a position where I want to seed to multiple peers that are getting NAT-ed behind a shared IP. The tracker correctly recognizes all the peers, but my seeding instance only seems to recognize one at a time.

Would love an option to change this behavior.

### Which application of Transmission?

transmission-daemon (also seen in Mac client)

### Which version of Transmission?

3.00 and 2.94

**Comment by himabindugit** [comment]
@ckerr any comments for this issue?

**Comment by ckerr** [comment]
Just to clarify, is this request about serving a single torrent to multiple peers on the same address? 

Transmission has no problem serving multiple torrents to the same peer; I'm doing that right now on my seedbox...

**Comment by benfrew13** [comment]
@ckerr correct. Multiple torrents to the same peer, no issue. Multiple peers on the same address (but different ports), only seeing one peer in the list on the seeder and only a single transfer at a time.

**Comment by ckerr** [comment]
This would require a medium-sized refactor to peer-mgr, which currently uses torrent + address as a unique key. This would need to be changed to be torrent + address + port.

It's a reasonable idea, but IDK how many people this will effect and it's not going to be high on my TODO list. If anyone wants to write a PR for this I'd be happy to answer questions and do code review. A good starting point would probably be to change

```diff
-- get_existing_atom(tr_address const& addr)
++ get_existing_atom(tr_address const& addr, tr_port port)
```

in peer-mgr.cc and then just keep working backwards from there until all the compile errors are fixed.

**Comment by krazak** [comment]
Hello @ckerr,
The changes were made in [here](https://github.com/krazak/transmission).  The branch is [feature/allow-simultaneous-downloads-multiple-peers-with-same-address](https://github.com/krazak/transmission/tree/feature/allow-simultaneous-downloads-multiple-peers-with-same-address).  I'm able to build this fine with XCode, however in the terminal, I get the following error:

`make[2]: *** No rule to make target "/Users/krazak/dev/transmission/qt/-silent', needed by "qt/-silent.qm".  Stop.
make[1]: *** [qt/CMakeFiles/transmission-qt.dir/all] Error 2
make: *** [all] Error 2`

Also, I'm not able to find instructions on how to build the rpm package.  Could you provide some guidance or documentation that I can follow so I can test this change.  Thanks!

**Comment by ckerr** [comment]
@krazak just open up a PR and mark it as a draft; that way CI can do all the work for you

**Comment by GaryElshaw** [comment]
If you follow the link to the overview page, you can drill down further. Be warned, it might take some time ;-)

**Comment by krazak** [comment]
@ckerr I have been unsuccessful when trying to build the RPM package for CentOS 7.  I get the following error:
![image](https://user-images.githubusercontent.com/5414213/214769900-5fa8ca94-b7e1-440b-9523-81f687565573.png)
The problem is that to resolve this issue, version 7.81.0 of curl needs to be installed.  When I use `yum install libcurl-devel` the curl version that is installed is 7.29.0.  I have not found a later version than 7.29.0 for CentOS 7.  Do you have any suggestions on how to get passed this error?  Thanks!

**Comment by ckerr** [comment]
@krazak please open a new issue for this and I'll fix it; we ought to build cleanly with 7.29.0
